[BE] Fix MPS build warnings

malfet opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #83049
* __->__ #83048

Mostly get rid of unused variables, but also:
- Hide `-[MPSGraphTensorData printNDArray]` behind undefined method
  access
- Rename several methods in ScatterGather/TriangularOps:
   - `scatterAlongAxisWithDataTensor:`->`scatterAlongAxis:withDataTensor:`
   - `gatherAlongAxisWithUpdatesTensor:`->`gatherAlongAxis:withUpdatesTensor:`
   - `getCoordinateValueWithShapeTensor:`->`coordinateAlongAxisTensor:withShapeTensor:` 
- Add `-Wno-unguarded-availability-new` to suppress 12.3+ availability warnings

malfet committed:
Make MPS compile warning free

Mostly get rid of unused variables, but also:
- Hide `-[MPSGraphTensorData printNDArray]` behind undefined method
  access

[ghstack-poisoned]

malfet committed:
Update on "[BE] Make MPS compile warning free"


Mostly get rid of unused variables, but also:
- Hide `-[MPSGraphTensorData printNDArray]` behind undefined method
  access
- Add `MPSGraph+CoolMethods.h` that contains definitions of undefined methods used by MPS ops

[ghstack-poisoned]

malfet committed:
Update on "[BE] Make MPS compile warning free"


Mostly get rid of unused variables, but also:
- Hide `-[MPSGraphTensorData printNDArray]` behind undefined method
  access
- Add `MPSGraph+CoolMethods.h` that contains definitions of undefined methods used by MPS ops

[ghstack-poisoned]

kulinseth commented on aten/src/ATen/native/mps/MPSGraph+CoolMethods.h:
```diff
@@ -0,0 +1,30 @@
+                                                axis:(NSInteger)dim
+                                                name:(NSString * _Nullable)name;
+
+-(MPSGraphTensor *) getCoordinateValueWithShapeTensor:(MPSGraphTensor *) shapeTensor
```
@malfet, can you please replace these methods with the Public SDK ones, also please remove this file. We won't need CoolMethods 😄 .
```
-(MPSGraphTensor *) coordinateAlongAxisTensor:(MPSGraphTensor *) axisTensor
                              withShapeTensor:(MPSGraphTensor *) shapeTensor
                                         name:(NSString * _Nullable) name
```
Note the axisTensor and shapeTensor are swapped.

kulinseth commented on aten/src/ATen/native/mps/MPSGraph+CoolMethods.h:
```diff
@@ -0,0 +1,30 @@
+// TODO: Do not call to undocumented methods
+
+@interface MPSGraph(CoolMethods)
+-(MPSGraphTensor *) gatherAlongAxisWithUpdatesTensor:(MPSGraphTensor *) updateTensor
```
For this the the right API is:

```
-(MPSGraphTensor *) gatherAlongAxis:(NSInteger) axis
                  withUpdatesTensor:(MPSGraphTensor *) updatesTensor
                      indicesTensor:(MPSGraphTensor *) indicesTensor
                               name:(NSString * _Nullable) name
```

kulinseth commented on aten/src/ATen/native/mps/MPSGraph+CoolMethods.h:
```diff
@@ -0,0 +1,30 @@
+                                           axisTensor:(MPSGraphTensor *) axisTensor
+                                                 name:(NSString * _Nullable) name;
+
+-(MPSGraphTensor *) scatterAlongAxisWithDataTensor:(MPSGraphTensor *) dataTensor
```
For this the right API is:
```
-(MPSGraphTensor *) scatterAlongAxis:(NSInteger) axis
                      withDataTensor:(MPSGraphTensor *) dataTensor
                       updatesTensor:(MPSGraphTensor *) updatesTensor
                       indicesTensor:(MPSGraphTensor *) indicesTensor
                                mode:(MPSGraphScatterMode) mode
                                name:(NSString * _Nullable) name
```

malfet committed:
Update on "[BE] Fix MPS build warnings"


Mostly get rid of unused variables, but also:
- Hide `-[MPSGraphTensorData printNDArray]` behind undefined method
  access
- Add `MPSGraph+CoolMethods.h` that contains definitions of undefined methods used by MPS ops
- Add `-Wno-unguarded-availability-new` to suppress 12.3+ availability warnings

[ghstack-poisoned]

malfet commented:
@kulinseth updated selector names per your suggestions

albanD commented on aten/src/ATen/mps/MPSGuardImpl.mm:
```diff
@@ -9,9 +9,6 @@
   void MPSGuardImpl::createEvent(
     mpsEvent_t* event,
     const EventFlag flag) const {
-    id<MTLDevice> mtl_device = MPSDevice::getInstance()->device();
```
Wait what is this function doing now? Both args are ignored and it is a no-op??

malfet replied:
I don't know what its doing, but even before the change it was a no-op, wasn't it?
But I guess it was supposed to do something like `*event = new mpsEvent(flag);`?

albanD replied:
It was indeed a noop before so this is ok. But we should fix this for sure!

albanD reviewed (approved):
Very nice cleanup.
Thanks for taking the time to do this Nikita!
